lib/SILOptimizer/Analysis/CallerAnalysis.cpp

```diff
@@ -11,64 +11,175 @@
+            callerInfo.calleeStates.insert(calleeFn);
+            auto &optMin = iter.first->second.numPartialAppliedArguments;
+            unsigned min = optMin.getValueOr(UINT_MAX);
+            for (ApplySite partialSite : result.partialApplySites) {
+              min = std::min(min, partialSite.getNumArguments());
+            }
+            optMin = min;
+#ifndef NDEBUG
+            for (auto applySite : result.partialApplySites) {
+              visitedCallSites.insert(applySite.getInstruction());
+              callSitesThatMustBeVisited.remove(applySite.getInstruction());
+            }
+#endif
+          }
           continue;
+        }
+      }
 
-        // Update the callee information for this function.
-        FunctionInfo &CallerInfo = FuncInfos[F];
-        CallerInfo.Callees.insert(CalleeFn);
-        
-        // Update the callsite information for the callee.
-        FunctionInfo &CalleeInfo = FuncInfos[CalleeFn];
-        CalleeInfo.Callers.insert(F);
+#ifndef NDEBUG
+      // Make sure that we are in sync with FullApplySite.
+      if (auto apply = FullApplySite::isa(&i)) {
+        if (apply.getCalleeFunction() && !visitedCallSites.count(&i)) {
+          callSitesThatMustBeVisited.insert(&i);
+        }
         continue;
       }
-      if (auto *PAI = dyn_cast<PartialApplyInst>(&II)) {
-        SILFunction *CalleeFn = PAI->getCalleeFunction();
-        if (!CalleeFn)
-          continue;
 
-        // Update the callee information for this function.
-        FunctionInfo &CallerInfo = FuncInfos[F];
-        CallerInfo.Callees.insert(CalleeFn);
-        
-        // Update the partial-apply information for the callee.
-        FunctionInfo &CalleeInfo = FuncInfos[CalleeFn];
-        int &minAppliedArgs = CalleeInfo.PartialAppliers[F];
-        int numArgs = (int)PAI->getNumArguments();
-        if (minAppliedArgs == 0 || numArgs < minAppliedArgs) {
-          minAppliedArgs = numArgs;
+      // Make sure that we are in sync with looking for partial apply callees.
+      if (auto *pai = dyn_cast<PartialApplyInst>(&i)) {
+        if (pai->getCalleeFunction() && !visitedCallSites.count(&i)) {
+          callSitesThatMustBeVisited.insert(pai);
         }
         continue;
       }
+#endif
     }
   }
+
+#ifndef NDEBUG
+  if (callSitesThatMustBeVisited.empty())
+    return;
+  llvm::errs() << "Found unhandled call sites!\n";
+  while (callSitesThatMustBeVisited.size()) {
+    auto *i = callSitesThatMustBeVisited.pop_back_val();
+    llvm::errs() << "Inst: " << *i;
+  }
+  assert(false && "Unhandled call site?!");
+#endif
 }
 
-void CallerAnalysis::invalidateExistingCalleeRelation(SILFunction *F) {
-  FunctionInfo &CallerInfo = FuncInfos[F];
-  for (auto Callee : CallerInfo.Callees) {
-    FunctionInfo &CalleeInfo = FuncInfos[Callee];
-    CalleeInfo.Callers.erase(F);
-    CalleeInfo.PartialAppliers.erase(F);
+void CallerAnalysis::invalidateExistingCalleeRelation(SILFunction *caller) {
```

Comment:
I don't know what `Existing` implies here. If you don't want to refer to the computed callees simply as `callees`, then I would use the term "known callees" consistently:
`calleeStates` -> `knownCallees`
`invalidateExistingCalleeRelation` -> `invalidateKnownCallees`